tests/unit/einstein/test_ceph.py

```diff
@@ -0,0 +1,162 @@
+        the image Raises
+        """
+        self.assertRaises(file_system_exceptions.ArgumentsOutOfRangeException,
+                          self.fs.read, CEPH_IMG, RANGE_OUTSIDE_THE_IMAGE,
+                          len(TEST_DATA))
+        self.assertRaises(file_system_exceptions.ArgumentsOutOfRangeException,
+                          self.fs.write, CEPH_IMG, TEST_DATA,
+                          RANGE_OUTSIDE_THE_IMAGE)
+        self.assertRaises(file_system_exceptions.ArgumentsOutOfRangeException,
+                          self.fs.discard, CEPH_IMG, RANGE_OUTSIDE_THE_IMAGE,
+                          len(TEST_DATA))
+
+    def test_write(self):
+        """Test write image"""
+        original_data = self.fs.read(CEPH_IMG, OFFSET, len(TEST_DATA))
+        bytes_written = self.fs.write(CEPH_IMG, TEST_DATA, OFFSET)
+        data_read = self.fs.read(CEPH_IMG, OFFSET, bytes_written)
+        self.assertEqual(TEST_DATA, data_read)
+        self.fs.discard(CEPH_IMG, OFFSET, bytes_written)
+        data_after_read = self.fs.read(CEPH_IMG, OFFSET, bytes_written)
+        self.assertEqual(original_data, data_after_read)
+
+    def tearDown(self):
+        self.fs.remove(CEPH_IMG)
+
+
+class TestSnapshot(unittest.TestCase):
+    """Test snapshot operations"""
+    @trace
+    def setUp(self):
+        self.fs = ceph.RBD(_cfg.fs, _cfg.iscsi.password)
+        self.fs.create_image(CEPH_IMG, CEPH_IMG_SIZE)
+        self.fs.snap_image(CEPH_IMG, CEPH_SNAP_IMG)
+
+    def test_snap_image(self):
+        """Test create snapshot"""
+        snapList = self.fs.list_snapshots(CEPH_IMG)
+        self.assertIn(CEPH_SNAP_IMG, snapList)
+
+    def test_image_exist_exception(self):
+        """this exception will be raised if snap image already exists"""
+        self.assertRaises(file_system_exceptions.ImageExistsException,
+                          self.fs.snap_image, CEPH_IMG, CEPH_SNAP_IMG)
+
+    def snap_protect(self):
+        """protecting a snapshot"""
+        self.fs.snap_protect(CEPH_IMG, CEPH_SNAP_IMG)
+        isProtected = self.fs.snap_is_protected(CEPH_IMG, CEPH_SNAP_IMG)
+        self.assertTrue(isProtected, "Failed to protect a snapshot!")
+
+    def test_clone(self):
+        """Test clone image"""
+        self.snap_protect()
+        cloneFlag = self.fs.clone(CEPH_IMG, CEPH_SNAP_IMG,
+                                  CEPH_CHILD_IMG)
+        self.assertTrue(cloneFlag, "Failed to clone image")
+        children = self.fs.list_children(CEPH_IMG, CEPH_SNAP_IMG)
+        children_imgs = [child[1] for child in children if child[0] == POOL]
+        self.assertIn(CEPH_CHILD_IMG, children_imgs,
+                      "Error, cannot find child image")
+        parent_info = self.fs.get_parent_info(CEPH_CHILD_IMG)
+        self.assertEqual(CEPH_IMG, parent_info[1])
+        self.assertEqual(CEPH_SNAP_IMG, parent_info[2])
+        self.fs.flatten(CEPH_CHILD_IMG)
+        self.snap_unprotect()
```

Comment:
the flatten and unprotect should be part of tearDown. Again you don't need to flatten the clone, you can just remove it.